File: Shapes7/Circle.hs

-- (C) 2004-2005, Oleg Kiselyov & Ralf Laemmel
-- Haskell's overlooked object system

module Circle where

import Shape


-- The composed type of circles

data CircleData =
     CircleData { getShape :: ShapeData
                 , getRadius :: Int
                 }


-- A "closed" constructor

circle x y r
 = CircleData { getShape  = shape x y
              , getRadius = r
              }


-- A circle is a shape

instance Shape CircleData
 where
  moveTo x y s = s { getShape = moveTo' x y (getShape s) }
  rMoveTo deltax deltay s = s { getShape = rMoveTo' deltax deltay (getShape s) }
  draw s
    =  putStrLn ("Drawing a Circle at:("
    ++ (show (getX (getShape s)))
    ++ ","
    ++ (show (getY (getShape s)))
    ++ "), radius "
    ++ (show (getRadius s)))

File: java/InterfaceOriented/Circle.java
class Circle implements IShape {
    private Shape base;
    private int radius;

    // constructor
    Circle(int newx, int newy, int newradius) {
	base = new Shape(newx,newy);
	setRadius(newradius);
    }

    // accessors for the radius
    int getRadius() { return radius; }
    void setRadius(int newradius) { radius = newradius; }

    // forward move messages
    public void moveTo(int newx, int newy) {
	base.moveTo(newx,newy); 
    }
    public void rMoveTo(int deltax, int deltay) { 
	base.rMoveTo(deltax,deltay); 
    }

    // draw the circle
    public void draw() {
	System.out.println(
	   "Drawing a Circle at:(" + base.getX() + ", " + base.getY() +
           "), radius " + getRadius());
    }
}

interface ICircle extends IShape {
    int getRadius();
    void setRadius(int newradius);
}
